src/kernel/outwards.jl

```diff
@@ -32,21 +32,37 @@ is occupied.
 
 Pass grid name `Symbol`s to `R` and `W` type parameters to use specific grids.
 """
+# Updated the OutwardsDispersal rule to include an optional mask
 struct OutwardsDispersal{R,W,S<:Stencils.AbstractKernelStencil} <: SetNeighborhoodRule{R,W}
     stencil::S
+    mask::M
 end
+
+function OutwardsDispersal{R,W}(stencil::S; mask=nothing) where {R,W,S<:Stencils.AbstractKernelStencil}
+    OutwardsDispersal{R,W,S}(stencil, mask)
+end
+
 function OutwardsDispersal{R,W}(; kw...) where {R,W}
     OutwardsDispersal{R,W}(DispersalKernel(; kw...))
 end
 
 @inline function applyrule!(data, rule::OutwardsDispersal{R,W}, N, I) where {R,W}
     N == zero(N) && return nothing
+    # Check if the current cell is masked, skip if it is
+    if rule.mask !== nothing && !rule.mask[I...]
+        return nothing
+    end
     sum = zero(N)
     for (offset, k) in zip(offsets(rule), kernel(rule))
-        @inbounds propagules = N * k
-        @inbounds add!(data[W], propagules, I .+ offset...)
-        sum += propagules
+        target = I .+ offset
+        # Check if the target cell is within bounds and not masked
+        (target_mod, inbounds) = _inbounds(Reflect(), size(data[1]), target...)
```

Comment:
I think `inbounds(data, i`)` does exactly this for you

Reply:
True! And it's better cause I don't need to specify the boundary condition

Reply:
Yes we try to keep anything related to boundary conditions, aux data, masks etc out of rules and in the `data` object. Then we can swap them around without changing all our code.